inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/TableMetricStatementExecutor.java

```diff
@@ -0,0 +1,172 @@
+    private final SinkMetricData sinkMetricData;
+    private final AtomicInteger counter = new AtomicInteger();
+    private transient FieldNamedPreparedStatement st;
+
+    /**
+     * Keep in mind object reuse: if it's on then key extractor may be required to return new
+     * object.
+     */
+    public TableMetricStatementExecutor(StatementFactory stmtFactory, JdbcRowConverter converter,
+            DirtySinkHelper<Object> dirtySinkHelper, SinkMetricData sinkMetricData) {
+        this.stmtFactory = checkNotNull(stmtFactory);
+        this.converter = checkNotNull(converter);
+        this.batch = new ArrayList<>();
+        this.dirtySinkHelper = dirtySinkHelper;
+        this.sinkMetricData = sinkMetricData;
+    }
+
+    /**
+     * parses an SQL exception message, and returns dirty records
+     *
+     * @param e the exception
+     * @return a string that can be compared with the return value of other parseRecord calls
+     */
+    public static void parseRecord(Exception e, List<Integer> answer) {
+        final Pattern pattern = Pattern.compile("Batch entry (\\d+) ");
+        Matcher matcher = pattern.matcher(e.getMessage());
+        if (matcher.find()) {
+            answer.add(Integer.parseInt(matcher.group(1)));
+        }
+        // if e is sql exciption, identify all dirty data, else identify only one dirty data.
+        if (e instanceof SQLException) {
+            SQLException next = ((SQLException) e).getNextException();
+            if (next != null) {
+                parseRecord(next, answer);
+            }
+        }
+    }
+
+    @Override
+    public void prepareStatements(Connection connection) throws SQLException {
+        st = stmtFactory.createStatement(connection);
+    }
+
+    @Override
+    public void addToBatch(RowData record) throws SQLException {
+        batch.add(record);
+        converter.toExternal(record, st);
+        st.addBatch();
+    }
+
+    @Override
+    public void executeBatch() throws SQLException {
+        try {
+            st.executeBatch();
+            addMetrics();
+        } catch (SQLException e) {
+            if (counter.incrementAndGet() == 3) {
+                // parse record from error and handle
+                List<Integer> dirtyRecords = new ArrayList<>();
+                parseRecord(e, dirtyRecords);
+                handleDirty(dirtyRecords);
+                counter.set(0);
+            } else {
+                throw new SQLException(e);
+            }
+        }
+    }
+
+    private void addMetrics() {
+        Set<RowData> set = new HashSet<>();
+        int rowCount = 0;
+        long rowSize = 0;
+        for (RowData record : batch) {
+            if (!set.contains(record)) {
+                set.add(record);
+                rowCount++;
+                rowSize += record.toString().getBytes(StandardCharsets.UTF_8).length;
+            }
+        }
+        sinkMetricData.invoke(rowCount, rowSize * 8);
```

Comment:
`*8` why?

Reply:
deleted... but the rowsize will /8  for successful data, possibly due byte[].length.